### PR TITLE
Muon - LoadAndApplyMuonDetectorGrouping algorithm

### LINKED ISSUE #22506: Muon - LoadAndApplyMuonDetectorGrouping
In the grouping tab of muon analysis the behaviour of the Load grouping button will be replicated by creating a new algorithm. Below is where to find the current code.

Load - https://github.com/mantidproject/mantid/blob/master/Framework/API/src/GroupingLoader.cpp

The muon analysis GUI currently has functions for reading/writing XML files containing pair/group information from the table. These can be re-purposed into the new algorithm via a helper class.

This algorithm will make use of the recently created ApplyMuonDetectorGrouping algorithm.

### PR DESCRIPTION
**Pull request by TWJubb**
closes #22506

Description of work
-------------------

A new algorithm for muon analysis, which loads grouping information from an XML file (which is created through the save grouping button in the grouping tab of the muon analysis GUI) and applies it to some data, saving the results to the ADS.

The algorithm uses two recently created algorithms #22510 and #22452.

Testing
-------

**Setup**

Open the Muon Analysis GUI (`Interface->Muon->Muon Analysis`) and load MUSR00015190.nxs from the usage data. Close the GUI. Delete all the workspaces in the ADS except "MuonAnalysis".

[ExampleGrouping.zip](https://github.com/mantidproject/mantid/files/2116103/ExampleGrouping.zip)

**Test 1**

Open the algorithm from `Muon->Data Handling-> LoadAndApplyMuonDetectorGrouping`. Set **filename** to be the XML file attached to this PR. Set the **InputWorkspace** to MuonAnalysis and leave everything else blank. Leave **ApplyAsymmetryToGroups** unchecked.

When run, the workspace group "MUSR00015190" should be created and filled with workspaces. Several other workspaces/tables may be created but ignore those.

Now check **ApplyAsymmetryToGroups** and run again, more workspaces should appear in addition to those already in "MUSR00015190".

**Test 2**

As with Test 1, but modify any of the analysis options : 

- Setting **RebinArgs** should affect only workspaces with names not ending in "_Raw".
- Setting **TimeOffset** should affect all workspaces.
- Setting **SummedPeriods** to "1,2" should update the data, but **does not** update the names (this will be fixed), similarly with **SubtractedPeriods**.

Fixes #22506

Does this update require release notes?
- [ ] Yes
- [ ] No

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Muon/inc/MantidMuon/LoadAndApplyMuonDetectorGrouping.h
Comment (AnthonyLim23) [inline Framework/Muon/inc/MantidMuon/LoadAndApplyMuonDetectorGrouping.h]: I thought we had added the pairing algorithm

File: Framework/Muon/src/LoadAndApplyMuonDetectorGrouping.cpp
Comment (AnthonyLim23) [inline Framework/Muon/src/LoadAndApplyMuonDetectorGrouping.cpp]: micro seconds

File: Framework/Muon/src/LoadAndApplyMuonDetectorGrouping.cpp
Comment (AnthonyLim23) [inline Framework/Muon/src/LoadAndApplyMuonDetectorGrouping.cpp]: can we just say that the name is the run number (e.g. MUSR0006256)?

File: Framework/Muon/src/LoadAndApplyMuonDetectorGrouping.cpp
Comment (AnthonyLim23) [inline Framework/Muon/src/LoadAndApplyMuonDetectorGrouping.cpp]: I would suggest removing the if and just have the assignment (if's are slow)

File: Framework/Muon/src/LoadAndApplyMuonDetectorGrouping.cpp
Comment (AnthonyLim23) [inline Framework/Muon/src/LoadAndApplyMuonDetectorGrouping.cpp]: then you can place the next 2 lines in an else. That way everything uses the same return and its a bit neater

File: Framework/Muon/src/LoadAndApplyMuonDetectorGrouping.cpp
Comment (AnthonyLim23) [inline Framework/Muon/src/LoadAndApplyMuonDetectorGrouping.cpp]: We currently do not produce this sort of table. I think we should make it an option of the algorithm if to produce it or not.

File: Framework/Muon/src/LoadAndApplyMuonDetectorGrouping.cpp
Comment (AnthonyLim23) [inline Framework/Muon/src/LoadAndApplyMuonDetectorGrouping.cpp]: if this file exists then the other algorithms (load group/pair) should update it

File: Framework/Muon/src/LoadAndApplyMuonDetectorGrouping.cpp
Comment (AnthonyLim23) [inline Framework/Muon/src/LoadAndApplyMuonDetectorGrouping.cpp]: you probably also want to store the run number, so you know which group corresponds to which run

File: Framework/Muon/src/LoadAndApplyMuonDetectorGrouping.cpp
Comment (AnthonyLim23) [inline Framework/Muon/src/LoadAndApplyMuonDetectorGrouping.cpp]: lets add the group name to the error message. So then the user knows what is invalid. The same with the below

File: Framework/Muon/src/LoadAndApplyMuonDetectorGrouping.cpp
Comment (AnthonyLim23) [inline Framework/Muon/src/LoadAndApplyMuonDetectorGrouping.cpp]: lets use the explicit expressions above in the if statements

File: Framework/Muon/src/LoadAndApplyMuonDetectorGrouping.cpp
Comment (AnthonyLim23) [inline Framework/Muon/src/LoadAndApplyMuonDetectorGrouping.cpp]: why is this called DTC?

**Review by AnthonyLim23 — CHANGES_REQUESTED** [review]
I think all of the XML related code should be in its own file on its own. The we only need to change one file to update the saving/loading. The save XML method seems to only be used for testing, so that should be in some sort of XML testing helper

File: Framework/Muon/src/LoadAndApplyMuonDetectorGrouping.cpp
Comment (TWJubb) [reply Framework/Muon/src/LoadAndApplyMuonDetectorGrouping.cpp]: Removing the if means that the `ads.retreive(wsName)` on the next line would throw an error. So it's either the if statement or a try/catch. Unless there is a third option I'm not aware of.

File: Framework/Muon/src/LoadAndApplyMuonDetectorGrouping.cpp
Comment (TWJubb) [reply Framework/Muon/src/LoadAndApplyMuonDetectorGrouping.cpp]: I think it's best to hold off on these modifications for now. I've added the option to produce the table and defaulted it to false.

If this table needs to be used during scripting then there is a substantial amount of functionality that would be added which is probably best for another PR; e.g. having to update the table, store multiple tables for multiple runs, updating the other muon grouping/pairing algorithms.

**Comment by TWJubb** [comment]
I will move the XML code in the next PR, after I have merged with the other algorithms.

**Comment by AnthonyLim23** [comment]
you need to add this to the release notes (and I just noticed that the pair and grouping algorithms also need adding).

File: docs/source/release/v3.13.0/muon.rst
Comment (AnthonyLim23) [inline docs/source/release/v3.13.0/muon.rst]: add that the input is an xml file

File: Framework/Muon/src/MuonAlgorithmHelper.cpp
Comment (AnthonyLim23) [inline Framework/Muon/src/MuonAlgorithmHelper.cpp]: delete the code instead of just commenting it out

**Comment by AnthonyLim23** [comment]
After a little work I can get the results to match the muon analysis. However, the unnormalized workspaces are missing and need adding

**Comment by AnthonyLim23** [comment]
https://github.com/mantidproject/mantid/pull/22653/files#diff-d7969b02fe958c6da198112b8d6f39b4R474

**Comment by TWJubb** [comment]
The unNorm workspaces are now added in the Asymmetry calculation for groups, with associated unit tests. I used the opportunity to neaten up the tests for the `ApplyMuonDetectorGrouping` algorithm and add tests for the `MuonWorkspaceCreationHelper`.

File: Framework/Muon/src/ApplyMuonDetectorGrouping.cpp
Comment (AnthonyLim23) [inline Framework/Muon/src/ApplyMuonDetectorGrouping.cpp]: we can define `"_unNorm` at the top of the file in an anonymous namespace https://github.com/mantidproject/mantid/pull/22653/files#diff-ef8bcda06d8d7c7d120651e19b290d73R68

File: Framework/Muon/test/ApplyMuonDetectorGroupingTest.h
Comment (AnthonyLim23) [inline Framework/Muon/test/ApplyMuonDetectorGroupingTest.h]: I don't understand this as I cannot see where setup goes.

File: Framework/Muon/test/ApplyMuonDetectorGroupingTest.h
Comment (AnthonyLim23) [inline Framework/Muon/test/ApplyMuonDetectorGroupingTest.h]: why have these numbers changed?

File: Framework/Muon/test/ApplyMuonDetectorGroupingTest.h
Comment (AnthonyLim23) [inline Framework/Muon/test/ApplyMuonDetectorGroupingTest.h]: delete this function or give it something to do

File: Framework/Muon/test/ApplyMuonDetectorGroupingTest.h
Comment (TWJubb) [reply Framework/Muon/test/ApplyMuonDetectorGroupingTest.h]: The generated 'fake' data has changed, for one thing we no longer allow negative counts.

File: Framework/Muon/test/ApplyMuonDetectorGroupingTest.h
Comment (TWJubb) [reply Framework/Muon/test/ApplyMuonDetectorGroupingTest.h]: What do you mean?

Comment (AnthonyLim23) [reply Framework/Muon/test/ApplyMuonDetectorGroupingTest.h]: to me it looks like the object is setUpADSWithWorkspace  and the function is setup

Comment (TWJubb) [reply Framework/Muon/test/ApplyMuonDetectorGroupingTest.h]: The object is `setUpADSWithWorkspace` (defined in namespace at the top), the instance is called `setup` which takes the workspace as an argument and puts it in the ADS, as well as performing a few other common tasks for testing the algorithm.

File: Framework/Muon/src/LoadAndApplyMuonDetectorGrouping.cpp
Comment (AnthonyLim23) [reply Framework/Muon/src/LoadAndApplyMuonDetectorGrouping.cpp]: can we have an issue for this?

File: Framework/Muon/src/ApplyMuonDetectorGrouping.cpp
Comment (AnthonyLim23) [inline Framework/Muon/src/ApplyMuonDetectorGrouping.cpp]: the convention is `name+unnorm+raw`, you currently have `name+raw+unnorm`

**Review by AnthonyLim23 — APPROVED** [review]

**Review by AnthonyLim23 — CHANGES_REQUESTED** [review]
Just noticed that the release notes need to be moved from 3.13 to 3.14

File: docs/source/release/v3.13.0/muon.rst
Comment (AnthonyLim23) [inline docs/source/release/v3.13.0/muon.rst]: need to delete this line as it will cause errors

**Comment by AnthonyLim23** [comment]
when I try to use it I get the error ```Invalid value for TimeMin: Time values are negative.```

File: docs/source/release/v3.13.0/muon.rst
Comment (TWJubb) [reply docs/source/release/v3.13.0/muon.rst]: What errors would it cause? The builds all seemed to work ok. Should all double white lines be removed in these files?

Comment (AnthonyLim23) [reply docs/source/release/v3.13.0/muon.rst]: you need a single line after the bullets

**Comment by TWJubb** [comment]
I have added the ability to crop workspaces from the algorithm. This is added to bring the algorithm closer to the functionality of the MuonAnalysis GUI.

# Time Values in the Muon GUI
---------------------------

In the GUI Home Tab there are two `lineEdit`s "Time Zero" and "First Good Data". The values for both of these can be loaded from the nexus file if required. Internally, the GUI defines a `TimeZero`, `LoadedTimeZero` and `FirstGoodData`. 

## Time Zero

**When the GUI starts**, all the workspace x-values (times in mu s) are shifted by `LoadedTimeZero`. 

Any subsequent shift, caused by the user changing the values in some part of the GUI, is given by the equation

```
time shift = LoadedTimeZero - TimeZero
```

So, if you want to shift the time axis **forward**, you must enter a **negative** time into the "Time Zero" box in the Home Tab. In this algorithm, you would instead simple add a **positive** value into the "TimeOffset" property. Thus we can replicate the behaviour of the GUI, with a little care to ensure we enter the correct combination of times into "TimeOffset"

## First Good Data

The behaviour of "First Good Data" is different, in that it **crops** the workspace. Thus the Load algorithm is unable to replicate the behaviour of a first good data in the file (one cannot simply crop after running the algorithm as several calculations performed in the algorithm are not commutative with the cropping, e.g. the asymmetry).

Adding the ability to crop the workspace allows us to replicate the "First Good Data" value by simply entering it in the "TimeMin" property.

**Comment by AnthonyLim23** [comment]
can we have a way that will only crop the beginning and not the end?

**Comment by TWJubb** [comment]
Group Asymmetry
---------------

During the normal usage of the muon analysis GUI, a user may want to apply an asymmetry calculation (to groups) restricted to a particular range. They may not necessarily want to crop the data to this range though, thus we need a few extra inputs to separate the asymmetry range from the cropping range.

**Comment by AnthonyLim23** [comment]
you can accept the asymmetry values but you never use them

**Review by AnthonyLim23 — APPROVED** [review]
Once  the merge conflict has been fixed it is good to go

**Pull request merged**
